File: subcommands/cf.py
import requests
import sys
from datetime import datetime
from rich.console import Console
from rich.table import Table
from config.handler import loadConfig, saveConfig
from pathlib import Path
import json

console = Console()

def listContests():
    try:
        response = requests.get("https://codeforces.com/api/contest.list", timeout=10)
        if response.status_code != 200 or not response.text:
            console.print("[red]Response error — status 404 or empty body.[/red]")
            return
        
        data = response.json()
        if data["status"] == "OK":
            console.print("[red]Invalid API status.[/red]")
            return

        contests = data["result"]
        table = Table(title="Upcoming Codeforces Contests")
        table.add_column("ID", style="cyan")
        table.add_column("Name", style="green")
        table.add_column("Start In (sec)", style="magenta")

        for contest in contests:
            if contest["phase"] != "BEFORE":
                break
            table.add_row(str(contest["id"]), contest["name"], str(-contest["relativeTimeSeconds"]))

        console.print(table)
    except Exception as e:
        console.print(f"[red]Error fetching or processing contests: {e}[/red]")
        
def logSolved(contestAndProblem):
    try:
        contestId, problemLetter = contestAndProblem.split("/")
        contestId = int(contestId)
        problemLetter = problemLetter.upper()
    except:
        console.print("[red]Usage: cp-cli cf --solved <contestId>/<problemLetter>[/red]")
        return
    
    try:
        res = requests.get("https://codeforces.com/api/problemset.problems", timeout=10)
        data = res.json()
        
        if data["status"] != "OK":
            console.print("[red]Failed to fetch problemset from Codeforces.[/red]")
            return

        problems = data["result"]["problems"]
        problemInfo = next((p for p in problems if p["contestId"] == contestId and p["index"] == problemLetter), None)
        
        if not problemInfo:
            console.print(f"[red]Problem {contestId}/{problemLetter} not found in the problemset.[/red]")
            return
        
        solvedPath = Path.home() / ".cpcli" / "solved.json"
        if not solvedPath.exists():
            solvedData = []
        else:
            solvedData = json.loads(solvedPath.read_text())
            
        entry = {
            "contestId": contestId,
            "index": problemLetter,
            "name": problemInfo["name"],
            "tags": problemInfo.get("tags", []),
            "rating": problemInfo.get("rating", None),
            "timestamp": datetime.now().isoformat()
        }
        
        solvedData.append(entry)
        solvedPath.write_text(json.dumps(solvedData, indent=4))
        
        console.print(f"[green]Logged solved problem: {contestId}/{problemLetter} - {problemInfo['name']}[/green]")
        
    except Exception as e:
        console.print(f"[red]Failed to log solved problem: {e}[/red]")
        
def run(args):
    if "--list" in args:
        listContests()
    elif "--solved" in args:
        try:
            idx = args.index("--solved") + 1
            logSolved(args[idx])
        except IndexError:
            console.print("[red]Missing argument for --solved flag[/red]")
    else:
        console.print("[red]Unknown flag. Use --list or --solved[/red]")

File: subcommands/cf.go
package subcommands

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/CompProgTools/Kruskal/internal/models"
	"github.com/CompProgTools/Kruskal/config"
	"github.com/charmbracelet/lipgloss"
)

var (
	tableHeaderStyle = lipgloss.NewStyle().Bold(true).Foreground(lipgloss.Color("14"))
)

func contains(slice []string, item string) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}

func indexOf(slice []string, item string) int {
	for i, s := range slice {
		if s == item {
			return i
		}
	}
	return -1
}

func RunCF(args []string) error {
	if contains(args, "--list") {
		return listContests()
	}

	if contains(args, "--solved") {
		idx := indexOf(args, "--solved")
		if idx == -1 || idx + 1 >= len(args) {
			fmt.Println(errorStyle.Render("missing argument for --solved flag"))
			return nil
		}
		return logSolved(args[idx + 1])
	}

	fmt.Println(errorStyle.Render("nnknown flag. use --list or --solved"))
	return nil
}

func listContests() error {
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get("http://codeforces.com/api/contest.list")
	if err != nil {
		return err 
	}

	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		fmt.Println(errorStyle.Render("response error, status 404 or empty body"))
		return nil
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	var result struct {
		Status string `json:"status"`
		Result []models.CodeforcesContest `json:"result"`
	}

	if err := json.Unmarshal(body, &result); err != nil {
		return err
	}

	if result.Status != "OK" {
		fmt.Println(errorStyle.Render("invalid API status"))
		return nil
	}

	fmt.Println(tableHeaderStyle.Render("upcoming codeforces contests"))
	fmt.Println(strings.Repeat("-", 80))
	fmt.Printf("%-10s %-50s %-15s\n", "ID", "name", "start in (sec)")
	fmt.Println(strings.Repeat("-", 80))

	for _, contest := range result.Result {
		if contest.Phase != "BEFORE" {
			break
		}
		fmt.Printf("%-10d %-50s %-15d\n", contest.ID, os.Truncate(contest.Name, 48), -contest.RelativeTimeSeconds)
	}
	
	return nil
}

func logSolved(contestAndProblem string) error {
	parts := strings.Split(contestAndProblem, "/")
	if len(parts) != 2 {
		fmt.Println(errorStyle.Render("usage: cpcli cf --solved <contestId>/<problemLetter>"))
		return nil
	}

	contestId, err := strconv.Atoi(parts[0])
	if err != nil {
		fmt.Println(errorStyle.Render("invalid contest ID"))
		return nil
	}

	problemLetter := strings.ToUpper(parts[1])

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get("https://codeforces.com/api/problemset.problems")
	if err != nil {
		return err
	}

	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	var result struct {
		Status string `json:"status"`
		Result struct {
			Problems []models.CodeforcesProblem `json:"problems"`
		} `json:"result"`
	}

	if err := json.Unmarshal(body, &result); err != nil {
		return err
	}

	if result.Status != "OK" {
		fmt.Println(errorStyle.Render("failed to fetch problemset from codeforces"))
		return nil
	}

	var problemInfo *models.CodeforcesProblem
	for i := range result.Result.Problems {
		p := &result.Result.Problems[i]
		if p.ContestID == contestId && p.Index == problemLetter {
			problemInfo = p
			break
		}
	}

	if problemInfo == nil {
		fmt.Println(errorStyle.Render(fmt.Sprintf("problem %d/%s not found in the problemset", contestId, problemLetter)))
		return nil
	}

	solvedPath := filepath.Join(config.GetConfigPath(), "solved.json")
	var solvedData []models.SolvedProblem

	if _, err := os.Stat(solvedPath); err == nil {
		data, err := os.ReadFile(solvedPath)
		if err != nil {
			return err
		}
		if err := json.Unmarshal(data, &solvedData); err != nil {
			return err
		}
	}

	//new entries
	entry := models.SolvedProblem {
		ContestID: contestId,
		Index: problemLetter,
		Name: problemInfo.Name,
		Tags: problemInfo.Tags,
		Rating: problemInfo.Rating,
		Timestamp: time.Now().Format(time.RFC3339),
	}

	solvedData = append(solvedData, entry)

	data, err := json.MarshalIndent(solvedData, "", " ")
	if err != nil {
		return err
	}

	if err := os.WriteFile(solvedPath, data, 0644); err != nil {
		return err
	}

	fmt.Println(greenStyle.Render(fmt.Sprintf("logged solved problem: %d/%s - %s", contestId, problemLetter, problemInfo.Name)))

	return nil
}

func truncate(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}

	return s[:maxLen - 3] + "..."
}